src/map/pc.cpp

```diff
@@ -15235,6 +15235,68 @@ void pc_attendance_claim_reward( struct map_session_data* sd ){
 	clif_attendence_response( sd, attendance_counter );
 }
 
+/**
+ * Send a player to jail and determine the location to send in jail.
+ * @param sd: Player data
+ * @param duration: Duration in minutes (default INT_MAX = infinite)
+ */
+void pc_jail(map_session_data &sd, int32 duration) {
+	uint16 m_index = mapindex_name2id(MAP_JAIL);
+	int32 x, y;
+
+	switch (rnd() % 2) { // Jail Locations
+		case 0: // Default jail
+			x = 24;
+			y = 75;
+			break;
+		default: // Jail #1
+			x = 49;
+			y = 75;
+			break;
+	}
+
+	sc_start4(nullptr, &sd.bl, SC_JAILED, 100, duration, m_index, x, y, (duration != INT_MAX && duration != 0 ? 60000 : 1000));
+}
+
+/**
+ * Determine the punishment type when failing macro checks.
+ * @param sd: Player data
+ * @param stype: Macro detection status type (for banning)
+ */
+static void pc_macro_punishment(map_session_data &sd, e_macro_detect_status stype) {
+	int32 duration = 0;
+
+	// Determine if there's a unique duration
+	if (battle_config.macro_detection_punishment_time > 0) {
+		char time[13];
+
+		safesnprintf(time, 13, "+%dnm", battle_config.macro_detection_punishment_time);
+		duration = static_cast<int32>(solve_time(time));
+	}
+
+	if (battle_config.macro_detection_punishment == 0) { // Ban
+		clif_macro_detector_status(sd, stype);
+		chrif_req_login_operation(sd.macro_detect.reporter_aid, sd.status.name, (duration == 0 ? CHRIF_OP_LOGIN_BLOCK : CHRIF_OP_LOGIN_BAN), duration, 0, 0);
+	} else { // Jail
+		// Delete the timer
+		if (sd.macro_detect.timer != INVALID_TIMER)
+			delete_timer(sd.macro_detect.timer, pc_macro_detector_timeout);
```

Comment:
Not sure if that block (including the = INVALID_TIMER assignment) should be moved out of the if, to erase the timer also if the player will be banned. Not sure if we get timer mismatches otherwise.